docs(tutorials): nvidia triton server (#2065)
scaleway/docs-content: tutorials/nvidia-triton/index.mdx

---
meta:
  title: Deploying models with NVIDIA Triton Inference Server on Scaleway Object Storage
  description: This page shows how to deploy models with NVIDIA Triton Inference Server on Scaleway Object Storage
content:
  h1: Deploying models with NVIDIA Triton Inference Server on Scaleway Object Storage
  paragraph: This page shows how to deploy models with NVIDIA Triton Inference Server on Scaleway Object Storage
categories:

>   - compute

Reviewer: ```suggestion
  - compute
  - observability
```

    ./fetch_models.sh
    ```
5. Navigate to the `server/docs/examples/model_repository` directory within the cloned repository.
6. Upload the example model folder to your bucket in Scaleway Object Storage. You can use the [Scaleway Object Storage API](/docs/storage/object/api-cli/using-api-call-list/), any S3 compatible tool, or web interface to upload the model folder.
    <Message type="tip">
      You can use the `s3cmd` [command-line tool](/docs/tutorials/s3cmd/) or any other S3 compatible tool to upload your data.
    </Message>

## Configuring Triton Inference Server

1. Log into your GPU Instance using SSH.
    ```
    ssh root@<gpu_instance_ip>
    ```
3. Edit the Triton server configuration file, typically named `config.pbtxt`. This file should be located in the Triton Inference Server installation directory.
4. Configure the model repository path by adding or modifying the `model-repository` parameter:
   ```bash
   model-repository: "s3://https://s3.fr-par.scw.cloud:443/<your-bucket-name>/<path-to-model-folder>"
   ```
   Replace `<your-bucket-name>` and `<path-to-model-folder>` with the appropriate values for your Scaleway Object Storage setup.
   <Message type="tip">

>      When using S3, the credentials and default region can be passed by using either the [created a AWS-CLI configuration file](/storage/object/api-cli/object-storage-aws-cli/) or via the respective environment variables. If the environment variables are set they will take a higher priority and will be used by Triton instead of the credentials set using the AWS-CLI configuration file.

Reviewer: ```suggestion
     When using S3, the credentials and default region can be passed by using either the [created AWS-CLI configuration file](/storage/object/api-cli/object-storage-aws-cli/) or via the respective environment variables. If the environment variables are set they will take a higher priority and will be used by Triton instead of the credentials set using the AWS-CLI configuration file.
```

   </Message>
5. Save the configuration file.

## Starting Triton Inference Server

1. Pull the Triton Inference Server container image:
   ```bash
   docker pull nvcr.io/nvidia/tritonserver:23.07-py3
   ```
   <Message type="tip">
     In case Docker is not preinstalled on your GPU Instance, follow the [Docker installation guide](/tutorials/install-docker-ubuntu-jammy-jellyfish/) before pulling the image.
   </Message>

2. Start the Triton Inference Server container:
   ```bash
   docker run --gpus all --rm -p8000:8000 -p8001:8001 -p8002:8002 \
   -v<path_to_your_model_repository>:/models nvcr.io/nvidia/tritonserver:23.07-py3> \
   tritonserver --model-repository=s3://https://s3.fr-par.scw.cloud:443/<your_bucket_name>/<path-to-model-folder>
   ```
   - `--gpus all`: Utilizes all available GPUs for inference (if present).

>    - `-p8000:8000 -p8001:8001 -p8002:8002`: Exposes ports for HTTP, GRPC, and Metrics.

Reviewer: ```suggestion
   - `-p8000:8000 -p8001:8001 -p8002:8002`: Exposes ports for HTTP, GRPC, and metrics.
```

   - `-v<path_to_your_model_repository>:/models`: Mounts your local model repository inside the container.
   - `--model-repository=s3://https://s3.fr-par.scw.cloud:443/<your-bucket-name>/<path-to-model-folder>`: The path to your Scaleway Object Storage bucket

    Once Triton Server is running, you can access it for inference using the appropriate REST APIs or SDKs provided by NVIDIA.

> ## Exporting metrics to Scaleway observability Cockpit

Reviewer: ```suggestion
## Exporting metrics to Scaleway Observability Cockpit
```

> Triton Inference Server provides metrics that can be exported for monitoring. Let's set up Prometheus and Scaleway Cockpit to visualize these metrics.

Reviewer: ```suggestion
Triton Inference Server provides metrics that can be exported for monitoring. In this section, we will set up Prometheus and Scaleway Cockpit to visualize these metrics.
```

Configure Triton to export metrics in Prometheus format by adding the following lines to the Triton server configuration:
   ```ini
   metrics: [
     {
       name: "prometheus"
       kind: PROMETHEUS
       output: "https://<SECRET_KEY>:<COCKPIT_API_KEY>@metrics.cockpit.fr-par.scw.cloud/api/v1/push"
     }
   ]
   ```
   Replace `<SECRET_KEY>` and `<COCKPIT_API_KEY>` with your Cockpit [token credentials](/observability/cockpit/how-to/create-token/).

> Your metrics are now pushed to cockpit and you can [access Grafana](/observability/cockpit/how-to/access-grafana-and-managed-dashboards/) to visualize them.

Reviewer: ```suggestion
Your metrics are now pushed to Cockpit and you can [access Grafana](/observability/cockpit/how-to/access-grafana-and-managed-dashboards/) to visualize them.
```
